### ISSUE TITLE
panic on arm64

### ISSUE THREAD

**Issue by uschen** [cockroachdb/pebble#952]
```
[5058]: unexpected fault address 0x6c1ed94dad07336
[5058]: fatal error: fault
[5058]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x6c1ed94dad07336 pc=0x601e70]
[5058]: goroutine 30 [running]:
[5058]: runtime.throw(0x98160e, 0x5)
[5058]:         /usr/local/go/src/runtime/panic.go:1116 +0x54 fp=0x400060ed80 sp=0x400060ed50 pc=0x413d4
[5058]: runtime.sigpanic()
[5058]:         /usr/local/go/src/runtime/signal_unix.go:727 +0x3b8 fp=0x400060edb0 sp=0x400060ed80 pc=0x56a88
[5058]: github.com/cockroachdb/pebble/internal/cache.(*entry).linkFile(...)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/entry.go:120
[5058]: github.com/cockroachdb/pebble/internal/cache.(*shard).metaAdd(0x4000158d70, 0x2, 0xa, 0x93c04, 0x4000523ec0, 0x4000523ec0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:310 +0x140 fp=0x400060ee00 sp=0x400060edc0 pc=0x601e70
[5058]: github.com/cockroachdb/pebble/internal/cache.(*shard).Set(0x4000158d70, 0x2, 0xa, 0x93c04, 0x4000136600, 0x0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:141 +0x528 fp=0x400060eea0 sp=0x400060ee00 pc=0x6015d8
[5058]: github.com/cockroachdb/pebble/internal/cache.(*Cache).Set(0x40004aa6e0, 0x2, 0xa, 0x93c04, 0x4000136600, 0x931)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:658 +0x6c fp=0x400060eee0 sp=0x400060eea0 pc=0x602ffc
[5058]: github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0x400040ea00, 0x93c04, 0x92c, 0x0, 0x400017d030, 0x5, 0x24, 0x40004fa000)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:1519 +0x220 fp=0x400060f0f0 sp=0x400060eee0 pc=0x617750
[5058]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock(0x400017ce00, 0x400017cef8)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:212 +0x124 fp=0x400060f150 sp=0x400060f0f0 pc=0x612c04
[5058]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).skipForward(0x400017ce00, 0x0, 0x0, 0x0, 0x0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:422 +0x44 fp=0x400060f1c0 sp=0x400060f150 pc=0x613bf4
[5058]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).Next(0x400017ce00, 0x66062c, 0x4000083380, 0x879, 0x400060f2e8)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:398 +0xf8 fp=0x400060f230 sp=0x400060f1c0 pc=0x613a08
[5058]: github.com/cockroachdb/pebble/sstable.(*twoLevelCompactionIterator).Next(0x400000dea0, 0x61eb98, 0x0, 0xc, 0x400060f2e8)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:913 +0x50 fp=0x400060f290 sp=0x400060f230 pc=0x616730
[5058]: github.com/cockroachdb/pebble.(*mergingIter).nextEntry(0x40004c4180, 0x4000083380)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/merging_iter.go:495 +0x6c fp=0x400060f2f0 sp=0x400060f290 pc=0x66053c
[5058]: github.com/cockroachdb/pebble.(*mergingIter).Next(0x40004c4180, 0x40000a29a8, 0x400060f418, 0x61d650, 0x40000a29a8)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/merging_iter.go:981 +0x64 fp=0x400060f330 sp=0x400060f2f0 pc=0x662474
[5058]: github.com/cockroachdb/pebble.(*compactionIter).iterNext(...)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:392
[5058]: github.com/cockroachdb/pebble.(*compactionIter).nextInStripe(0x40000de000, 0x0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:409 +0x34 fp=0x400060f390 sp=0x400060f330 pc=0x63e2d4
[5058]: github.com/cockroachdb/pebble.(*compactionIter).skipInStripe(0x40000de000)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:377 +0x38 fp=0x400060f3b0 sp=0x400060f390 pc=0x63e268
[5058]: github.com/cockroachdb/pebble.(*compactionIter).Next(0x40000de000, 0x40004fa0c0, 0x24, 0x30, 0x1)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:247 +0x50 fp=0x400060f480 sp=0x400060f3b0 pc=0x63d610
[5058]: github.com/cockroachdb/pebble.(*DB).runCompaction(0x400004f000, 0x3, 0x40000ca900, 0xad33a0, 0x1094be0, 0x40003a40f0, 0x400041a108, 0x1, 0x1, 0x0, ...)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:2234 +0xde0 fp=0x400060fc20 sp=0x400060f480 pc=0x63b0a0
[5058]: github.com/cockroachdb/pebble.(*DB).compact1(0x400004f000, 0x40000ca900, 0x0, 0x0, 0x0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1744 +0x144 fp=0x400060fe80 sp=0x400060fc20 pc=0x639cd4
[5058]: github.com/cockroachdb/pebble.(*DB).compact.func1(0xaeaea0, 0x400009b050)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1705 +0x90 fp=0x400060fef0 sp=0x400060fe80 pc=0x675320
[5058]: runtime/pprof.Do(0xaeaea0, 0x400009b050, 0x40003b3600, 0x1, 0x1, 0x4000489798)
[5058]:         /usr/local/go/src/runtime/pprof/runtime.go:40 +0xac fp=0x400060ff60 sp=0x400060fef0 pc=0x605a5c
[5058]: github.com/cockroachdb/pebble.(*DB).compact(0x400004f000, 0x40000ca900, 0x0)
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1702 +0xa0 fp=0x400060ffc0 sp=0x400060ff60 pc=0x639b70
[5058]: runtime.goexit()
[5058]:         /usr/local/go/src/runtime/asm_arm64.s:1136 +0x4 fp=0x400060ffc0 sp=0x400060ffc0 pc=0x72c64
[5058]: created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker
[5058]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1499 +0x2e8
[5058]: goroutine 1 [select]:
[5058]: net/http.(*Transport).getConn(0x40003ec780, 0x40004a7000, 0x0, 0x4000097080, 0x5, 0x4000039020, 0x1d, 0x0, 0x0, 0x0, ...)

```

**Comment by uschen** [comment]
```
[5051]: runtime: marked free object in span 0xffff916ea070, elemsize=96 freeindex=0 (bad use of unsafe.Pointer? try -d=checkptr)
[5051]: 0x4000592000 alloc marked
[5051]: 0x4000592060 free  unmarked
[5051]: 0x40005920c0 free  unmarked
[5051]: 0x4000592120 alloc marked
[5051]: 0x4000592180 alloc marked
[5051]: 0x40005921e0 free  unmarked
[5051]: 0x4000592240 alloc marked
[5051]: 0x40005922a0 free  unmarked
[5051]: 0x4000592300 free  unmarked
[5051]: 0x4000592360 alloc marked
[5051]: 0x40005923c0 alloc marked
[5051]: 0x4000592420 alloc marked
[5051]: 0x4000592480 free  unmarked
[5051]: 0x40005924e0 alloc marked
```

**Comment by petermattis** [comment]
@uschen This is not a type of crash I've seen before. Have you tried doing what the report suggests and compiling with `-d=checkptr`?

The Pebble cache does use `unsafe.Pointer` and manually managed memory. It is possible there is something fouled up there, though we haven't seen anything like this crash on Linux amd64.

**Comment by moredure** [comment]
https://github.com/cockroachdb/pebble/issues/954

**Comment by uschen** [comment]
unfortunately, build on arm64 (ubuntu 20) still panic (with clean db dir)

```
[17648]: unexpected fault address 0x516e394f476781
[17648]: fatal error: fault
[17648]: [signal SIGBUS: bus error code=0x1 addr=0x516e394f476781 pc=0x601648]
[17648]: goroutine 40 [running]:
[17648]: runtime.throw(0x981813, 0x5)
[17648]:         /usr/local/go/src/runtime/panic.go:1116 +0x54 fp=0x4000742490 sp=0x4000742460 pc=0x413d4
[17648]: runtime.sigpanic()
[17648]:         /usr/local/go/src/runtime/signal_unix.go:717 +0x420 fp=0x40007424c0 sp=0x4000742490 pc=0x56af0
[17648]: github.com/cockroachdb/pebble/internal/cache.(*refcnt).acquire(...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/refcnt_normal.go:26
[17648]: github.com/cockroachdb/pebble/internal/cache.(*Value).acquire(...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/value.go:39
[17648]: github.com/cockroachdb/pebble/internal/cache.(*entry).acquireValue(...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/entry.go:152
[17648]: github.com/cockroachdb/pebble/internal/cache.(*shard).Get(0x4000159108, 0x2, 0x9, 0x6e578, 0x4000159108)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:111 +0xa8 fp=0x4000742520 sp=0x40007424d0 pc=0x601648
[17648]: github.com/cockroachdb/pebble/internal/cache.(*Cache).Get(0x40003a4640, 0x2, 0x9, 0x6e578, 0x4000110000)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:650 +0x64 fp=0x4000742560 sp=0x4000742520 pc=0x603684
[17648]: github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0x40000c0a00, 0x6e578, 0x3a, 0x0, 0x0, 0x94b340, 0x9, 0x400069b180)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:1421 +0x4c fp=0x4000742770 sp=0x4000742560 pc=0x617c9c
[17648]: github.com/cockroachdb/pebble/sstable.(*Reader).readIndex(...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:1406
[17648]: github.com/cockroachdb/pebble/sstable.(*twoLevelIterator).init(0x400069b180, 0x40000c0a00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:636 +0x4c fp=0x40007427c0 sp=0x4000742770 pc=0x6154cc
[17648]: github.com/cockroachdb/pebble/sstable.(*Reader).NewIter(0x40000c0a00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:1346 +0x98 fp=0x4000742830 sp=0x40007427c0 pc=0x6177a8
[17648]: github.com/cockroachdb/pebble.(*tableCacheShard).newIters(0x40000c5100, 0x4000124a80, 0x400000e4c8, 0x1, 0x1, 0x0, 0x1, 0x4000139a38, 0x0, 0x0, ...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/table_cache.go:204 +0x204 fp=0x40007428e0 sp=0x4000742830 pc=0x66ca44
[17648]: github.com/cockroachdb/pebble.(*tableCache).newIters(0x400004f0c0, 0x4000124a80, 0x400000e4c8, 0x1, 0x1, 0x0, 0x1, 0x4000139a38, 0x0, 0x0, ...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/table_cache.go:56 +0x74 fp=0x4000742970 sp=0x40007428e0 pc=0x66c074
[17648]: github.com/cockroachdb/pebble.(*tableCache).newIters-fm(0x4000124a80, 0x400000e4c8, 0x1, 0x1, 0x0, 0x1, 0x4000139a38, 0x0, 0x0, 0x0, ...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/table_cache.go:53 +0x54 fp=0x4000742a00 sp=0x4000742970 pc=0x685044
[17648]: github.com/cockroachdb/pebble.(*levelIter).loadFile(0x40001399f0, 0x4000124a80, 0x1, 0x30)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/level_iter.go:311 +0x2b8 fp=0x4000742b20 sp=0x4000742a00 pc=0x65cbc8
[17648]: github.com/cockroachdb/pebble.(*levelIter).SeekGE(0x40001399f0, 0x4000192750, 0x24, 0x30, 0x24, 0x24, 0x800, 0x8aa)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/level_iter.go:359 +0x74 fp=0x4000742b70 sp=0x4000742b20 pc=0x65cec4
[17648]: github.com/cockroachdb/pebble.(*getIter).Next(0x4000139950, 0xffff88f6c108, 0x4000080c00, 0xffff88f6c108, 0x0)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/get_iter.go:149 +0x744 fp=0x4000743000 sp=0x4000742b70 pc=0x650f14
[17648]: github.com/cockroachdb/pebble.(*getIter).First(0x4000139950, 0x645e68, 0x4000743088, 0x645e90, 0x400)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/get_iter.go:60 +0x28 fp=0x4000743040 sp=0x4000743000 pc=0x650748
[17648]: github.com/cockroachdb/pebble.(*Iterator).First(0x4000139800, 0x4000139800)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/iterator.go:414 +0x120 fp=0x4000743090 sp=0x4000743040 pc=0x657100
[17648]: github.com/cockroachdb/pebble.(*DB).getInternal(0x400004f000, 0x4000192750, 0x24, 0x30, 0x0, 0x0, 0x24, 0x8d9, 0x8d9, 0xad38a0, ...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/db.go:422 +0x260 fp=0x40007430e0 sp=0x4000743090 pc=0x646070
[17648]: github.com/cockroachdb/pebble.(*DB).Get(...)
[17648]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/db.go:365
...
[17648]: net/http.serverHandler.ServeHTTP(0x4000370000, 0xae85c0, 0x400069c460, 0x40000c4c00)
[17648]:         /usr/local/go/src/net/http/server.go:2843 +0xbc fp=0x4000743c10 sp=0x4000743be0 pc=0x29d80c
[17648]: net/http.(*conn).serve(0x40004f2000, 0xaeb300, 0x40004ce000)
[17648]:         /usr/local/go/src/net/http/server.go:1925 +0x6f4 fp=0x4000743fc0 sp=0x4000743c10 pc=0x299af4
[17648]: runtime.goexit()
[17648]:         /usr/local/go/src/runtime/asm_arm64.s:1136 +0x4 fp=0x4000743fc0 sp=0x4000743fc0 pc=0x72c64
[17648]: created by net/http.(*Server).Serve
[17648]:         /usr/local/go/src/net/http/server.go:2969 +0x2e8
[17648]: goroutine 1 [chan receive]:
```

**Comment by uschen** [comment]
same when build on linux arm64 + 

```
CGO_ENABLED=0 GOOS=linux go build -a --installsuffix cgo --ldflags $(LD_FLAGS) -gcflags=all=-d=checkptr
```

```
[21426]: panic: runtime error: invalid memory address or nil pointer dereference
[21426]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x61e0b0]
[21426]: goroutine 19 [running]:
[21426]: github.com/cockroachdb/pebble/internal/cache.(*entry).linkFile(...)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/entry.go:120
[21426]: github.com/cockroachdb/pebble/internal/cache.(*shard).metaAdd(0x4000237050, 0x2, 0x8, 0xc6a2d, 0x400018d8c0, 0x400018d8c0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:310 +0x140
[21426]: github.com/cockroachdb/pebble/internal/cache.(*shard).Set(0x4000237050, 0x2, 0x8, 0xc6a2d, 0x40004c4600, 0x0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:141 +0x550
[21426]: github.com/cockroachdb/pebble/internal/cache.(*Cache).Set(0x40001146e0, 0x2, 0x8, 0xc6a2d, 0x40004c4600, 0x9b3)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/internal/cache/clockpro.go:658 +0x6c
[21426]: github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0x4000192500, 0xc6a2d, 0x9ae, 0x0, 0x40005be230, 0x5, 0x24, 0x40003b2c30)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:1519 +0x220
[21426]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).loadBlock(0x40005be000, 0x40005be0f8)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:212 +0x124
[21426]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).skipForward(0x40005be000, 0x0, 0x0, 0x0, 0x0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:422 +0x44
[21426]: github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).Next(0x40005be000, 0x67e83c, 0x4000259040, 0x8ee, 0x400063f2e8)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:398 +0xf8
[21426]: github.com/cockroachdb/pebble/sstable.(*twoLevelCompactionIterator).Next(0x400009f580, 0x63cac8, 0x0, 0xb, 0x400063f2e8)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/sstable/reader.go:913 +0x50
[21426]: github.com/cockroachdb/pebble.(*mergingIter).nextEntry(0x400057e180, 0x4000259040)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/merging_iter.go:495 +0x6c
[21426]: github.com/cockroachdb/pebble.(*mergingIter).Next(0x400057e180, 0x40004cf7a8, 0x400063f418, 0x63b580, 0x40004cf7a8)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/merging_iter.go:981 +0x64
[21426]: github.com/cockroachdb/pebble.(*compactionIter).iterNext(...)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:392
[21426]: github.com/cockroachdb/pebble.(*compactionIter).nextInStripe(0x4000259180, 0x0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:409 +0x34
[21426]: github.com/cockroachdb/pebble.(*compactionIter).skipInStripe(0x4000259180)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:377 +0x38
[21426]: github.com/cockroachdb/pebble.(*compactionIter).Next(0x4000259180, 0x40003b2d50, 0x24, 0x30, 0x1)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction_iter.go:247 +0x50
[21426]: github.com/cockroachdb/pebble.(*DB).runCompaction(0x4000080c00, 0x3, 0x4000576240, 0xb0db60, 0x10e4e20, 0x40002c60a0, 0x40000341c8, 0x1, 0x1, 0x0, ...)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:2234 +0xde0
[21426]: github.com/cockroachdb/pebble.(*DB).compact1(0x4000080c00, 0x4000576240, 0x0, 0x0, 0x0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1744 +0x144
[21426]: github.com/cockroachdb/pebble.(*DB).compact.func1(0xb25a20, 0x4000091b30)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1705 +0x90
[21426]: runtime/pprof.Do(0xb25a20, 0x4000091b30, 0x4000497620, 0x1, 0x1, 0x40000aa798)
[21426]:         /usr/local/go/src/runtime/pprof/runtime.go:40 +0xac
[21426]: github.com/cockroachdb/pebble.(*DB).compact(0x4000080c00, 0x4000576240, 0x0)
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1702 +0xa0
[21426]: created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker
[21426]:         /go/pkg/mod/github.com/cockroachdb/pebble@v0.0.0-20200930153740-4e0eb739ab1a/compaction.go:1499 +0x334
```

**Comment by uschen** [comment]
the code is basically:

concurrent read (small concurrency)
```
s.db.Get(key)
```

with batch write:
```
var b = s.db.NewBatch()
for ... {
if err := b.Set(key, v, nil); err != nil {
			return Err(err)
		}
}
return s.db.Apply(b, pebble.Sync) // also tried NoSync
```

**Comment by petermattis** [comment]
@uschen If this reproduces readily, my advice would be to start instrumenting the code around creating and deleting entries. See `entryAllocNew` and `entryAllocFree`. Also instrument the code in `entry.linkFile`, outputting both the pointer `e` and the pointers `e.fileLink.next` and `e.fileLink.prev`.

**Comment by uschen** [comment]
@petermattis will try that.

**Comment by uschen** [comment]
some update...

still happening with go 1.15.3

however, only on raspberry pi 4 4gb, `ubuntu Linux 5.4.0-1021-raspi `

not on jetson xavier nx `Linux 4.9.140-tegra`

both are tested on the xfs.

**Comment by uschen** [comment]
it seems the issue is similar to https://github.com/cockroachdb/pebble/issues/678

caused by `CGO_ENABLED=0`

**Comment by ivanjaros** [comment]
same on win10 x64. it suddenly appeared out of nowhere. i too do not allow cgo in my code base.

**Comment by ivanjaros** [comment]
since there is no mention in readme, does pebble require cgo? i thought pebble was pure go rewrite of leveldb...

**Comment by petermattis** [comment]
Cgo is used to allocate memory for the block cache when it is available. If Cgo is disabled, Pebble should still work, but there will be increased Go GC pressure because the block cache memory is allocated from the Go heap. You can disable cgo by compiling with `CGO_ENABLED=0`.

I'm curious, what system are using where cgo appears to be enabled, but then crashes at runtime? That sounds like a broken Go toolchain.

**Comment by ivanjaros** [comment]
i am already running `CGO_ENABLED=0 go build`

**Comment by ivanjaros** [comment]
the C version of `manual` eveb has this comment in the `New()` function:
```
	//   Note: the current implementation has a bug. While Go code is permitted
	//   to write nil or a C pointer (but not a Go pointer) to C memory, the
	//   current implementation may sometimes cause a runtime error if the
	//   contents of the C memory appear to be a Go pointer. Therefore, avoid
	//   passing uninitialized C memory to Go code if the Go code is going to
	//   store pointer values in it. Zero out the memory in C before passing it
	//   to Go.
```

**Comment by petermattis** [comment]
If you're passing `CGO_ENABLED=0`, you should be using https://github.com/cockroachdb/pebble/blob/master/internal/manual/manual_nocgo.go.

**Comment by ivanjaros** [comment]
> 
> 
> If you're passing `CGO_ENABLED=0`, you should be using https://github.com/cockroachdb/pebble/blob/master/internal/manual/manual_nocgo.go.

well..yes, but does not help me in any way :D
the library keeps on crashing. i am not sure if this is separate issue, or it is the same as this issue or the one linked by uschen. either way there is something wrong happening under the hood.

**Comment by petermattis** [comment]
> any test i can run to gather some debug info for you?

This isn't something I'll be able to debug remotely. It is all very mysterious which makes me curious about it, but I can't do much unless I can reproduce it. Give me a crash on amd64 and it will be fixed quickly

**Comment by ivanjaros** [comment]
> Give me a crash on amd64 and it will be fixed quickly

As i have said, I am on Win10 64bit, intel cpu and as far as i know it uses amd64 instruction set

**Comment by ivanjaros** [comment]
i am certain i do not have the spectre&co firmware updates and am unsure about windows itself(updates and such). so that could be something to look into, maybe? i am just guessing, i have no idea.

**Comment by petermattis** [comment]
> As i have said, I am on Win10 64bit, intel cpu and as far as i know it uses amd64 instruction set

I guess I should have mentioned Linux or Mac as well. I don't have access to Win10.

**Comment by ivanjaros** [comment]
ok, np. it is what it is.

i have picked pebble for my new project because badger stopped working on windows but it seems pebble isn't either :D ... even after i tested it beforehand. so i guess my search for a working and fast kvdb continues.

**Comment by petermattis** [comment]
@ivanjaros We have CI for Windows set up on this repo. So it does work to some extent. I'm really not sure why it is blowing up for you, nor do I have great knowledge about Go on Windows.

**Comment by ivanjaros** [comment]
if you look above you will see that in my case the panic is preceded by call to signal_windows.go and people who posed before me have the same but with signal_unix.go so something wrong sit syscalls or something like that maybe? In both cases preceded by refcnt_normal.go and so on. So there is clear path that this problem occurs on.

**Comment by petermattis** [comment]
I'm pretty sure `signal_{unix,windows}.go` is showing up because that is where the signal handlers for segfault or located. The rest of the stacktraces indicate that we're not near a system call.

**Comment by ivanjaros** [comment]
could this be user issue maybe? like misuse of methods or wrong flow of data or something on my end?

**Comment by petermattis** [comment]
> could this be user issue maybe? like misuse of methods or wrong flow of data or something on my end?

Certainly possible, but unlikely. These cache data structures are not exposed to the user (you). Typically, the only misuse of methods that can result in badness in the cache is failure to close iterators or DBs which results in a memory leak. Crashes like this should not be possible via Pebble API misuse, though because we don't know the cause we can't completely eliminate that possibility.

If you have a reproducible way to cause this crash, can you narrow it down to a small bit of code that you can share?

**Comment by ivanjaros** [comment]
so far it happen only in real application. when i tried to write dedicated test it has not happened yet. but if i will find some use case i will post i here.

btw if pebble is used in cockroachdb these days, don't you have some tests that run on windows?

**Comment by petermattis** [comment]
> btw if pebble is used in cdb these days, don't you have some tests that run on windows?

Yes, and those tests aren't seeing this problem, just as the Pebble CI tests on Windows aren't seeing this problem.

> btw2: could this have something to do with the prefix scans that we've been talking about in other issue(s)? like there is something wrong with the lower bound and upper bound values and their user/parsing? because i use this a lot in my code base.

It is super hard to just look at code like this and see a problem. I need a fully working reproduction case. 

That said, are you closing the `prefixIterator`? I think that should only cause a memory leak, but it could possibly be related.

**Comment by ivanjaros** [comment]
heureka! no wasn't closing the iterator. i guess since i was working mostly with bolt lately i forgot to check if it needs to be closed. thanks!! :D

**Comment by petermattis** [comment]
Glad to hear that was the problem, but I'm befuddled as to why it led to a crash.

**Comment by ivanjaros** [comment]
aaaah, nope. i am still getting `unexpected fault address 0xffffffffffffffff` and such. looks like it could be coming from tx.Get(). I check the error and return immediately after the call, if no error is present, then i process the data and after that i will call the closer. but maybe the closer needs to be called even if there is error. i will experiment. but i don't see how the closer is not nil if there is error.

**Comment by ivanjaros** [comment]
do i have to close iterator on tx before i can use tx(get)?

**Comment by petermattis** [comment]
@ivanjaros You're asking me to debug something with massively incomplete information. I can't help you unless you provide a complete working reproduction.

**Comment by ivanjaros** [comment]
i understand. but this is all i have. maybe if more people encounter this in the futur we can figure it but right now i am the outlier so i will move to another db for time being. thanks for your time.

**Comment by ivanjaros** [comment]
i tested my code with ubuntu in WSL and i got the same issue. this was with cgo disabled. when i enabled cgo there was no issue. so the problem is not the platform, or my code. the problem is plainly the non-cgo version of manual library as i though from the beginning. as for what is the culprit, i have no idea. i don't have knowledge of pebble to even start guessing.

**Comment by petermattis** [comment]
@ivanjaros I can't help you unless you provide a complete working reproduction.

**Comment by ivanjaros** [comment]
i understand. i am just reporting what i am seeing, for future reference, that's all :)

btw, are you running some of your CI builds with CGO_ENABLED=0?

**Comment by uschen** [comment]
not sure if it is related to cgo regression:
https://github.com/golang/go/issues?q=milestone%3AGo1.15.4+label%3ACherryPickApproved

**Comment by petermattis** [comment]
I'm pretty sure that this is the same problem as described in #1010 and should be fixed by #1011. 

@ivanjaros I see you reached a similar conclusion as to the problem, but I never saw your comment because Github doesn't alert me when a comment is edited, only when it is added.
